### ISSUE TITLE
Broken OTA behavior after #2303

### ISSUE THREAD

**Issue by mmakaay** [esphome/issues#2455]
### The problem

The current dev branch doesn't handle OTA correctly on my ESP8266 devices. I didn't test ESP32 yet.

One issue was a simple code issue, for which I already filed a PR (already merged).
It caused an invalid "Passwords do not match!" error, that is now fixed.
See: https://github.com/esphome/esphome/pull/2363

After fixing this, I am getting the following behavior when trying OTA updates:

* After flashing through a serial connection, the device does boot up correctly.
* When trying an OTA update, the initial handshake works, but at the start of transferring the actual firmware bin file (after a slight delay where the device is waiting for input), I get an exception and the device reboots because of that. 
**Update:** Fixed in https://github.com/esphome/esphome/pull/2366
* On this boot, the device reports 300000 suspected unsuccessful boot attempts, logically pushing the device into safe mode instead of standard operational mode.
**Update:** Fixed in https://github.com/esphome/esphome/pull/2367

The issues are possibly related through an issue with global preferences, but I have to look into this a bit further to see if I'm right.

### Which version of ESPHome has the issue?

2021.10.0-dev

### What type of installation are you using?

pip

### What platform are you using?

ESP8266

### Board

D1 mini and ESP-01

### Component causing the issue

ota

### Anything in the logs that might be useful for us?

```txt
Here some logging with ** annotations:

** Device boots after serial flashing.

[23:52:21]SDK:2.2.2-dev(38a443e)/Core:2.7.3-3-g2843a5ac=20703003/lwIP:STABLE-2_1_2_RELEASE/glue:1.2-30-g92add50/BearSSL:5c771be
[23:52:21][I][logger:216]: Log initialized

** The unsuccessful boot attempts start out as 0, as expected.

[23:52:21][C][ota:527]: There have been 0 suspected unsuccessful boot attempts.
[23:52:21][I][app:029]: Running through setup()...
...
[23:52:25][V][wifi_esp8266:467]: Event: Connected ssid='FOKKE' bssid=FC:EC:DA:8A:B7:CD channel=6
[23:52:25][C][api:025]: Setting up Home Assistant API server...
...
[23:52:38][D][api:098]: Accepted 192.168.100.135
[23:52:38][V][api.connection:712]: Hello from client: 'Home Assistant 2021.9.7 (192.168.100.135)'
[23:52:38][D][api.connection:728]: Home Assistant 2021.9.7 (192.168.100.135): Connected successfully

** After the device has booted, I start an OTA update.

[23:53:32][D][ota:253]: Starting OTA Update from 192.168.100.10...
[23:53:32][V][ota:282]: OTA features is 0x00
[23:53:32][V][ota:298]: Auth: Nonce is fe4ba2c07e57dabbb8f3d2158e9724a5
[23:53:32][V][ota:318]: Auth: CNonce is a8cad4bdd9e724d4bcefabae367370e6
[23:53:32][V][ota:325]: Auth: Result is 2a05ff8d18822004d5be5f93606bd704
[23:53:32][V][ota:333]: Auth: Response is 2a05ff8d18822004d5be5f93606bd704
[23:53:32][V][ota:361]: OTA size is 469248 bytes
[23:53:32]sleep disable
[23:53:32][begin] roundedSize:       0x00073000 (471040)
[23:53:32][begin] updateEndAddress:  0x003FB000 (4173824)
[23:53:32][begin] currentSketchSize: 0x00073000 (471040)
[23:53:32][begin] _startAddress:     0x00388000 (3702784)
[23:53:32][begin] _currentAddress:   0x00388000 (3702784)
[23:53:32][begin] _size:             0x00072900 (469248)
[23:53:32][V][ota:379]: Update: Binary MD5 is 6b0de4c1bca02ebc722f2034e6e75bf0

** In the command line esphome tool, I see the following error.
   (don't mind the byte size difference, this was from a different .bin file)

INFO Successfully compiled program.
INFO Connecting to 192.168.100.166
INFO Uploading test-serialdebug/.pioenvs/test-serialdebug/firmware.bin (408112 bytes)
Uploading: [=                                                           ] 1%
ERROR Error sending data: [Errno 104] Connection reset by peer

** At this point, an exception is raised on the serial console. I decoded the stack.

[23:53:34]
[23:53:34]--------------- CUT HERE FOR EXCEPTION DECODER ---------------
[23:53:34]
[23:53:34]Soft WDT reset

0x4010049c: esphome::delay(unsigned int) at src/esphome/components/esp8266/core.cpp line 13
0x402127dc: esphome::ota::OTAComponent::writeall_(unsigned char const*, unsigned int) at src/esphome/components/ota/ota_component.cpp line 493
0x4022822e: UpdaterClass::setMD5(char const*) at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/Updater.cpp line 204
0x40212e64: esphome::ota::OTAComponent::handle_() at src/esphome/components/ota/ota_component.cpp line 389
0x40100ce8: ets_post(uint8, ETSSignal, ETSParam) at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/core_esp8266_main.cpp line 177
0x40234fd0: _svfprintf_r at /home/earle/src/esp-quick-toolchain/repo/newlib/newlib/libc/stdio/nano-vfprintf.c line 526
0x40100ce8: ets_post(uint8, ETSSignal, ETSParam) at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/core_esp8266_main.cpp line 177
0x402381f9: glue2esp_linkoutput at glue-esp/lwip-esp.c line 301
0x4022bc84: AsyncClient::add(char const*, unsigned int, unsigned char) at .piolibdeps/test-serialdebug/ESPAsyncTCP-esphome/src/ESPAsyncTCP.cpp line 413
0x40238486: new_linkoutput at glue-lwip/lwip-git.c line 260
0x402388a0: ethernet_output at netif/ethernet.c line 312
0x4023fa0c: ip4_output_if_opt_src at core/ipv4/ip4.c line 1764
0x4023ecec: etharp_output_to_arp_index at core/ipv4/etharp.c line 770
0x40101826: check_poison_neighbors(uint16_t) at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/umm_malloc/umm_local.c line 71
0x401018e5: umm_malloc_core(size_t) at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/umm_malloc/umm_malloc.cpp line 458
0x4023ef30: etharp_output_LWIP2 at core/ipv4/etharp.c line 885
0x40101675: get_poisoned(void*, size_t) at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/umm_malloc/umm_poison.c line 117
0x4023fa0c: ip4_output_if_opt_src at core/ipv4/ip4.c line 1764
0x401010b4: malloc(size_t) at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/heap.cpp line 232
0x402388f5: do_memp_malloc_pool at core/memp.c line 255
0x4023fa50: ip4_output_if_opt at core/ipv4/ip4.c line 1577
0x4023fa76: ip4_output_if at core/ipv4/ip4.c line 1550
0x40240187: ip_chksum_pseudo at core/inet_chksum.c line 395
0x4023d1ae: tcp_output at core/tcp_out.c line 1361
0x40100ce8: ets_post(uint8, ETSSignal, ETSParam) at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/core_esp8266_main.cpp line 177
0x40101826: check_poison_neighbors(uint16_t) at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/umm_malloc/umm_local.c line 71
0x4010158f: umm_free_core(void*) at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/umm_malloc/umm_malloc.cpp line 351
0x40101b24: umm_free(void*) at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/umm_malloc/umm_malloc.cpp line 398
0x40100ce8: ets_post(uint8, ETSSignal, ETSParam) at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/core_esp8266_main.cpp line 177
0x40202adc: esphome::api::APIPlaintextFrameHelper::try_read_frame_(esphome::api::APIPlaintextFrameHelper::ParsedFrame*) at src/esphome/components/api/api_frame_helper.cpp line 766
0x40100ce8: ets_post(uint8, ETSSignal, ETSParam) at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/core_esp8266_main.cpp line 177
0x40100de2: millis() at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/core_esp8266_wiring.cpp line 188
0x402195ec: esphome::wifi::WiFiComponent::wifi_sta_connect_status_() at src/esphome/components/wifi/wifi_component_esp8266.cpp line 586
0x40100de2: millis() at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/core_esp8266_wiring.cpp line 188
0x402130ac: esphome::ota::OTAComponent::loop() at src/esphome/components/ota/ota_component.cpp line 218
0x4022d6f4: esphome::Component::call_loop() at src/esphome/core/component.cpp line 66
0x4022d79c: esphome::Component::call() at src/esphome/core/component.cpp line 96
0x4021add5: esphome::Application::loop() at src/esphome/core/application.cpp line 71
0x4021d43c: loop() at src/main.cpp line 262
0x4022987b: loop_wrapper() at /home/mmakaay1/.platformio/packages/framework-arduinoespressif8266@3.20704.0/cores/esp8266/core_esp8266_main.cpp line 197

** After the exception the device reboots.

[23:53:35] ets Jan  8 2013,rst cause:2, boot mode:(3,6)

** and comes up with a funny counter telling me that there were 300000 failed boots!

[23:53:35][C][ota:527]: There have been 300000 suspected unsuccessful boot attempts.
[23:53:35][E][ota:532]: Boot loop detected. Proceeding to safe mode.

** After which the device goes into safe boot.

[23:53:39][W][ota:211]: Last Boot was an unhandled reset, will proceed to safe mode in -299998 restarts
```

**Comment by mmakaay** [comment]
The issue with 300000 failed reboots, is caused by the load() function in the preferences storage not working at all for ESP8266.
After adding some debug statements to the code, I could see that the counter was stored correctly ( 1 + checksum 1123309631), and could be read back from rtc, but that the read data were not stored in the bytes that had to be filled by the load method.
```
[02:48:48][W][esp8266.preferences:151]: XXXX byte 0 before load = 224
[02:48:48][W][esp8266.preferences:151]: XXXX byte 1 before load = 147
[02:48:48][W][esp8266.preferences:151]: XXXX byte 2 before load = 4
[02:48:48][W][esp8266.preferences:151]: XXXX byte 3 before load = 0
[02:48:48][E][esp8266.preferences:115]: XXXX Read rtc 0 = 1
[02:48:48][E][esp8266.preferences:115]: XXXX Read rtc 1 = 1123309631
[02:48:48][W][esp8266.preferences:168]: XXXX byte 0 after load = 224
[02:48:48][W][esp8266.preferences:168]: XXXX byte 1 after load = 147
[02:48:48][W][esp8266.preferences:168]: XXXX byte 2 after load = 4
[02:48:48][W][esp8266.preferences:168]: XXXX byte 3 after load = 0
[02:48:48][E][ota:562]: XXXX val after read = 300000
```
So the 300000 value has nothing to do with actual data as stored in rtc.

**Comment by mmakaay** [comment]
Thinks look good and my PR has been merged into dev. Closing this issue as solved.
